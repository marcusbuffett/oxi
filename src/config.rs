use clap::Parser;
use serde::{Deserialize, Serialize};
use std::sync::OnceLock;

pub const NUM_GLOBALS: usize = 7;
pub const LEGAL_MOVES: usize = 64 * 76;
// Per-position square features:
// - 12 piece one-hots (white/black 6 each)
// - 1 en passant
// - 1 castling right at this square
// - 6 attackers (white)
// - 6 attackers (black)
// - 1 legal move count normalized (moves / 20)
// - 1 pawn isolated
// - 1 pawn backward
// - 1 pawn doubled
// - 1 square control (kept as last per-position feature)
pub const FEATURES_PER_SQUARE_POSITION: usize = 31;
pub const PREVIOUS_POSITIONS: usize = 3;
pub const FEATURES_PER_TOKEN: usize = FEATURES_PER_SQUARE_POSITION * (PREVIOUS_POSITIONS + 1) + 1;

// Global config storage
static GLOBAL_CONFIG: OnceLock<Config> = OnceLock::new();

/// Unified configuration for OXI chess engine training and inference
#[derive(Debug, Clone, Serialize, Deserialize, Parser)]
pub struct Config {
    // === DATA AND RUNTIME ===
    /// Path to data (PGN directory, PGN file, or CSV file)
    #[arg(long)]
    pub data_path: Option<std::path::PathBuf>,

    #[arg(long)]
    pub max_samples: Option<usize>,

    #[arg(long)]
    pub timeout: Option<u64>,

    /// Resume training from a specific checkpoint index
    #[arg(long)]
    pub checkpoint: Option<usize>,

    /// Train/validation split ratio (e.g., 0.9 for 90% train, 10% validation)
    #[arg(long, default_value = "0.9")]
    pub train_ratio: f32,

    /// Batch size for training
    #[arg(long)]
    pub batch_size: Option<usize>,

    /// Physical batch size (for gradient accumulation)
    #[arg(long, default_value = "256")]
    pub physical_batch_size: usize,

    /// Random seed for reproducibility
    #[arg(long, default_value = "42")]
    pub seed: u64,

    /// Number of data loader workers
    #[arg(long, default_value = "4")]
    pub num_workers: usize,

    /// Number of training epochs
    #[arg(long, default_value = "10")]
    pub num_epochs: usize,

    /// Using NoamLrScheduler
    #[arg(long, default_value = "1.0")]
    pub learning_rate: f64,

    #[arg(long, default_value = "4000")]
    pub warmup: usize,

    #[arg(long)]
    pub learning_rate_min: Option<f64>,

    /// Number of validation samples
    #[arg(long)]
    pub validation_samples: Option<usize>,

    /// Weight for policy loss
    #[arg(long, default_value = "2.0")]
    pub policy_loss_weight: f32,

    /// Label smoothing applied to the policy targets (smoothed over legal moves only)
    #[arg(long, default_value = "0.05")]
    pub policy_label_smoothing: f32,

    /// Weight for value loss
    #[arg(long, default_value = "0.2")]
    pub value_loss_weight: f32,

    /// Weight for side info loss
    #[arg(long, default_value = "0.5")]
    pub side_info_loss_weight: f32,

    #[arg(long, default_value = "2.0")]
    pub time_usage_loss_weight: f32,

    /// Weight decay for optimizer
    #[arg(long, default_value = "0.01")]
    pub weight_decay: f64,

    /// Gradient clipping norm (0 to disable)
    #[arg(long, default_value = "1.0")]
    pub gradient_clip: f64,

    /// Embedding dimension for tokens
    #[arg(long, default_value = "384")]
    pub embed_dim: usize,

    /// Number of attention heads
    #[arg(long, default_value = "8")]
    pub num_heads: usize,

    /// Number of K/V head groups for grouped-query attention (GQA). Must divide num_heads.
    /// Set equal to num_heads to disable grouping; set to 1 for multi-query attention (MQA).
    #[arg(long)]
    pub kv_groups: Option<usize>,

    /// Number of transformer layers
    #[arg(long, default_value = "10")]
    pub num_layers: usize,

    /// ELO bins for player skill levels
    #[arg(
        long,
        value_delimiter = ',',
        default_value = "800,1000,1200,1400,1600,1800,2000,2200,2400,2600,2800"
    )]
    pub elo_bins: Vec<i32>,

    /// Embedding dimension for ELO
    #[arg(long, default_value = "256")]
    pub elo_embed_dim: usize,

    /// MLP hidden dimension ratio
    #[arg(long, default_value = "4.0")]
    pub mlp_ratio: f32,

    /// Only include positions with a single legal move
    #[arg(long)]
    pub single_legal_move_only: bool,

    /// Only include positions with a single legal move
    #[arg(long, default_value = "false")]
    pub disable_tui: bool,

    /// Only include positions that are checkmate
    #[arg(long)]
    pub checkmate_only: bool,

    /// Probability of logging individual items for debugging (0.0 to 1.0)
    #[arg(long, default_value = "0.01")]
    pub item_log_probability: f32,

    /// Disable Shaw-style relative positional representations in attention
    #[arg(long, default_value = "false")]
    pub disable_shaw_pr: bool,

    /// Disable Rotary Positional Embeddings (RoPE)
    #[arg(long, default_value = "false")]
    pub disable_rope: bool,

    #[arg(long, default_value = "1")]
    pub num_devices: usize,
}

impl Config {
    /// Create new config with explicit parameters for testing
    pub fn new(embed_dim: usize, num_heads: usize, num_layers: usize) -> Self {
        Self {
            embed_dim,
            num_heads,
            num_layers,
            ..Default::default()
        }
    }

    // === MODEL ARCHITECTURE ACCESSORS ===

    pub fn embed_dim(&self) -> usize {
        self.embed_dim
    }
    pub fn global_dim(&self) -> usize {
        12
    }
    pub fn non_global_dim(&self) -> usize {
        self.embed_dim - self.global_dim()
    }
    pub fn num_heads(&self) -> usize {
        self.num_heads
    }
    pub fn kv_groups(&self) -> usize {
        self.kv_groups.unwrap_or(self.num_heads)
    }
    pub fn num_layers(&self) -> usize {
        self.num_layers
    }
    pub fn seq_len(&self) -> usize {
        64
    }
    pub fn head_dim(&self) -> usize {
        self.embed_dim / self.num_heads
    }

    pub fn elo_bins(&self) -> &[i32] {
        &self.elo_bins
    }
    pub fn elo_embed_dim(&self) -> usize {
        self.elo_embed_dim
    }
    pub fn mlp_ratio(&self) -> f32 {
        self.mlp_ratio
    }

    pub fn checkpoint_interval(&self) -> usize {
        1000
    }

    pub fn disable_rope(&self) -> bool {
        self.disable_rope
    }

    // etc... (other accessors)
}

/// Set the global config (should be called once at startup)
#[allow(clippy::result_large_err)]
pub fn set_global_config(config: Config) -> Result<(), Config> {
    GLOBAL_CONFIG.set(config)
}

/// Get the global config, falling back to default if not set
pub fn get_global_config() -> &'static Config {
    GLOBAL_CONFIG.get().unwrap()
}

impl Default for Config {
    fn default() -> Self {
        Self {
            data_path: None,
            train_ratio: 0.9,
            batch_size: None,
            physical_batch_size: 8,
            seed: 42,
            num_workers: 4,
            num_epochs: 10,
            weight_decay: 0.0001,
            gradient_clip: 1.0,
            learning_rate: 0.001,
            validation_samples: None,
            policy_loss_weight: 1.0,
            policy_label_smoothing: 0.05,
            value_loss_weight: 1.0,
            side_info_loss_weight: 0.5,
            embed_dim: 256,
            num_heads: 4,
            kv_groups: None,
            num_layers: 6,
            elo_bins: (800i32..=2800i32).step_by(200).collect(),
            elo_embed_dim: 128,
            mlp_ratio: 4.0,
            max_samples: None,
            timeout: None,
            checkpoint: None,
            single_legal_move_only: false,
            checkmate_only: false,
            item_log_probability: 1.00,
            time_usage_loss_weight: 0.0,
            disable_tui: true,
            learning_rate_min: None,
            warmup: 4000,
            disable_shaw_pr: false,
            disable_rope: false,
            num_devices: 1,
        }
    }
}

// Legacy type aliases for backward compatibility during transition
pub type ModelConfig = Config;
pub type TrainingConfig = Config;
