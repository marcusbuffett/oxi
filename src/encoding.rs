use shakmaty::{Bitboard, Chess, Color, EnPassantMode, File, Position, Role, Square};

use crate::config::{FEATURES_PER_SQUARE_POSITION, FEATURES_PER_TOKEN, PREVIOUS_POSITIONS};

pub fn encode_position(pos: &Chess, previous_positions: &[Chess]) -> Vec<f32> {
    let mut tokens = vec![0f32; 64 * FEATURES_PER_TOKEN];

    let mut all_positions = vec![Some(pos)];
    all_positions.extend(previous_positions.iter().map(Some));
    assert!(all_positions.len() <= PREVIOUS_POSITIONS + 1);
    all_positions.resize(PREVIOUS_POSITIONS + 1, None);

    for square in Square::ALL {
        let mut feature_idx = 0;
        let square_idx = square as usize * FEATURES_PER_TOKEN;
        for pos in &all_positions {
            if let Some(pos) = pos {
                let ep_square = pos.ep_square(EnPassantMode::Legal);
                let board = pos.board();
                if let Some(piece) = board.piece_at(square) {
                    let piece_offset = piece.role as usize - 1;
                    let color_offset = match piece.color {
                        Color::White => 0,
                        Color::Black => 6,
                    };
                    tokens[square_idx + feature_idx + color_offset + piece_offset] = 1.0;
                }
                feature_idx += 12;

                if Some(square) == ep_square {
                    tokens[square_idx + feature_idx] = 1.0;
                }
                feature_idx += 1;

                let castling = pos.castles().castling_rights();
                if castling.contains(square) {
                    tokens[square_idx + feature_idx] = 1.0;
                }
                feature_idx += 1;

                for color in [Color::White, Color::Black] {
                    let mut attackers = pos.king_attackers(square, color, board.occupied());
                    while let Some(square) = attackers.pop_front() {
                        if let Some(piece) = board.piece_at(square) {
                            tokens[square_idx + feature_idx + piece.role as usize - 1] = 1.0;
                        }
                    }
                    feature_idx += 6;
                }

                // Number of legal moves for the piece on this square, normalized and clamped
                let legal_moves_norm = normalized_legal_moves_for_square(pos, square);
                tokens[square_idx + feature_idx] = legal_moves_norm;
                feature_idx += 1;

                // Pawn structure features
                let (isolated, backward, doubled) = pawn_structure_features(pos, square);
                tokens[square_idx + feature_idx] = if isolated { 1.0 } else { 0.0 };
                feature_idx += 1;
                tokens[square_idx + feature_idx] = if backward { 1.0 } else { 0.0 };
                feature_idx += 1;
                tokens[square_idx + feature_idx] = if doubled { 1.0 } else { 0.0 };
                feature_idx += 1;

                // Add square control value as the last feature for this position
                let control_value = calculate_square_control(pos, square);
                tokens[square_idx + feature_idx] = control_value;
                feature_idx += 1;
            } else {
                feature_idx += FEATURES_PER_SQUARE_POSITION;
            }
            assert!(
                feature_idx % FEATURES_PER_SQUARE_POSITION == 0,
                "feature_idx % FEATURES_PER_SQUARE_POSITION == 0 {} % {} = {}",
                feature_idx,
                FEATURES_PER_SQUARE_POSITION,
                feature_idx % FEATURES_PER_SQUARE_POSITION
            );
        }

        tokens[square_idx + feature_idx] = square.is_dark() as usize as f32;
        feature_idx += 1;

        if feature_idx != FEATURES_PER_TOKEN {
            assert!(
                feature_idx == FEATURES_PER_TOKEN,
                "feature_idx == FEATURES_PER_TOKEN {feature_idx} should be {FEATURES_PER_TOKEN}"
            );
        }
        // if square_idx == 0 {
        //     tracing::info!("Position board after a1: {:?}", pos.board());
        //     log_encoded_board(&tokens);
        // }
    }
    // tracing::info!("Position board: {:?}", pos.board());
    // log_encoded_board(&tokens);
    // panic!();
    tokens
}
// let en_passant_offset = 12;
// if let Some(ep_square) = pos.ep_square(EnPassantMode::Legal) {
//     let idx = square_to_index(ep_square);
//     planes[en_passant_offset * 64 + idx] = 1.0;
// }
//
// let castling_offset = 13;
// let castling = pos.castles();
//
// // Mark corners based on castling rights
// if castling.has(Color::White, shakmaty::CastlingSide::KingSide) {
//     planes[castling_offset * 64 + 7] = 1.0; // h1
// }
// if castling.has(Color::White, shakmaty::CastlingSide::QueenSide) {
//     planes[castling_offset * 64 + 0] = 1.0; // a1
// }
// if castling.has(Color::Black, shakmaty::CastlingSide::KingSide) {
//     planes[castling_offset * 64 + 63] = 1.0; // h8
// }
// if castling.has(Color::Black, shakmaty::CastlingSide::QueenSide) {
//     planes[castling_offset * 64 + 56] = 1.0; // a8
// }

/// Get ELO bin index for a given rating
pub fn get_elo_bin(elo: i32, bins: &[i32]) -> usize {
    bins.iter().position(|&bin| elo < bin).unwrap_or(bins.len())
}

/// Get the piece value for square control calculation
fn get_piece_value(role: Role) -> f32 {
    match role {
        Role::Pawn => 1.0,
        Role::Knight => 3.0,
        Role::Bishop => 3.0,
        Role::Rook => 5.0,
        Role::Queen => 9.0,
        Role::King => 10.0, // King is valuable but not as much as queen for control
    }
}

/// Calculate square control for a given position and square
/// Returns a value from 0.0 to 1.0 where:
/// - 1.0 = total white control
/// - 0.0 = total black control  
/// - 0.5 = equal control
pub fn calculate_square_control(pos: &Chess, square: Square) -> f32 {
    let board = pos.board();
    let occupied = board.occupied();

    let mut white_control = 0.0f32;
    let mut black_control = 0.0f32;
    for color in [Color::White, Color::Black] {
        let mut occupied_clone = occupied;
        let mut board = pos.board().clone();
        loop {
            let Some(attacker) = board
                .attacks_to(square, color, occupied_clone)
                .into_iter()
                .min_by_key(|sq| sq.distance(square))
            else {
                break;
            };
            occupied_clone.toggle(attacker);
            let piece = board.remove_piece_at(attacker).unwrap();
            let piece_value = get_piece_value(piece.role);
            let control_contribution = 1.0 / piece_value.powf(2.0);

            match color {
                Color::White => white_control += control_contribution,
                Color::Black => black_control += control_contribution,
            }
        }
    }

    let total_control = white_control + black_control;
    if total_control == 0.0 {
        0.5 // No control from either side, neutral
    } else {
        white_control / total_control
    }
}

/// Compute number of legal moves available for the piece on a given square, normalized by 20
fn normalized_legal_moves_for_square(pos: &Chess, square: Square) -> f32 {
    let board = pos.board();
    let Some(_piece) = board.piece_at(square) else {
        return 0.0;
    };
    let mut count = 0usize;
    for m in pos.legal_moves() {
        if let Some(from_sq) = m.from() {
            if from_sq == square {
                count += 1;
            }
        }
    }
    ((count as f32) / 20.0).clamp(0.0, 1.0)
}

/// Determine pawn structure features for a pawn on the given square
/// Returns (isolated, backward, doubled)
fn pawn_structure_features(pos: &Chess, square: Square) -> (bool, bool, bool) {
    let board = pos.board();
    let Some(piece) = board.piece_at(square) else {
        return (false, false, false);
    };
    if piece.role != Role::Pawn {
        return (false, false, false);
    }
    let friendly_pawns = board.pawns() & board.by_color(piece.color);

    let file_idx = square.file() as i32; // 0..7
    let rank_idx = square.rank() as i32; // 0..7

    // Isolated: no pawns on adjacent files (any rank) of the same color
    let mut has_adjacent_file_pawn_same_color = false;
    for adj_file in [file_idx - 1, file_idx + 1] {
        if adj_file < 0 || adj_file > 7 {
            continue;
        }
        if (friendly_pawns & Bitboard::from_file(File::new(adj_file as u32))).count() > 0 {
            has_adjacent_file_pawn_same_color = true;
            break;
        }
    }
    let isolated = !has_adjacent_file_pawn_same_color;

    // Doubled: another pawn of same color on same file (different rank)
    let has_same_file_pawn_same_color_other_rank =
        (friendly_pawns & Bitboard::from_file(File::new(file_idx as u32))).count() > 1;
    let doubled = has_same_file_pawn_same_color_other_rank;

    // Backward: no friendly pawn on adjacent files behind this pawn (towards own home rank)
    // Build mask of adjacent files and intersect with friendly pawns, then scan ranks
    let adj_files_mask = if file_idx > 0 && file_idx < 7 {
        Bitboard::from_file(File::new((file_idx - 1) as u32))
            | Bitboard::from_file(File::new((file_idx + 1) as u32))
    } else if file_idx > 0 {
        Bitboard::from_file(File::new((file_idx - 1) as u32))
    } else {
        Bitboard::from_file(File::new((file_idx + 1) as u32))
    };
    let adjacent_friendly_pawns = friendly_pawns & adj_files_mask;
    let mut found_support = false;
    for sq in adjacent_friendly_pawns.into_iter() {
        let r = sq.rank() as i32;
        match piece.color {
            Color::White => {
                if r < rank_idx {
                    found_support = true;
                    break;
                }
            }
            Color::Black => {
                if r > rank_idx {
                    found_support = true;
                    break;
                }
            }
        }
    }
    let backward = !found_support;

    (isolated, backward, doubled)
}

#[cfg(test)]
mod tests {
    use super::*;
    use shakmaty::fen::Fen;
    use std::str::FromStr;

    #[test]
    fn test_square_control_f6_and_d4() {
        // Test position: r1bq1rk1/pp2ppbp/3p2p1/3N2B1/8/1P1Q4/1P3PPP/R4RK1 b - - 1 15
        let fen = "7r/ppp2kpr/2n2p2/2bnp3/8/3P4/PPPN1PPP/R1B2RK1 w - - 2 17";
        let fen_parsed: Fen = fen.parse().expect("Valid FEN");
        let pos: Chess = fen_parsed
            .into_position(shakmaty::CastlingMode::Standard)
            .expect("Valid position");
        let mut formatted_output = String::new();
        for square in ["a3", "h2", "h1"] {
            let square = Square::from_str(square).expect("Valid square");
            let control = calculate_square_control(&pos, square);
            formatted_output += &format!("{}: {:.6}\n", square, control);
        }

        insta::assert_snapshot!(formatted_output);
    }
}
