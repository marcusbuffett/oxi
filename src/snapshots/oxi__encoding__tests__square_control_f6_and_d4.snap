---
source: src/encoding.rs
expression: snapshot_data
---
Position: r1bq1rk1/pp2ppbp/3p2p1/3N2B1/8/1P1Q4/1P3PPP/R4RK1 b - - 1 15
f6 control: 0.417108
d4 control: 0.377153
